### PR TITLE
Add first post about PureNix

### PR DESCRIPTION
**Pull request by cdepillabout**

### CODE REVIEW

File: 2021-12-26-purenix.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+3.  Under the Hood of PureNix (coming soon)
+
+    This post explains how PureNix works internally.  It talks about what goes
+    in to writing a new PureScript backend.
+
+4.  PureNix vs. Nix (coming soon)
+
+    This post talks about when you should choose PureNix over writing in raw
+    Nix.
+
```
Comment (jonascarpay) [inline 2021-12-26-purenix.md]: What's the difference between 2 and 4? They seem like they could be the same post, no?

File: 2021-12-27-getting-started-with-purenix.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+identity a = a
+
+data Maybe a = Nothing | Just a
+
+fromMaybe :: forall a. a -> Maybe a -> a
+fromMaybe a Nothing = a
+fromMaybe _ (Just a) = a
+```
+
+PureNix compiles this PureScript code to the following Nix code[^1]:
```
Comment (jonascarpay) [inline 2021-12-27-getting-started-with-purenix.md]: FWIW, footnotes can be named, and the body doesn't have to be at the bottom: https://github.com/jonascarpay/jonascarpay.github.io/blob/master/posts/2021-05-16-parsing-programming-language.md?plain=1#L147

File: 2021-12-27-getting-started-with-purenix.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+## Using PureNix Professionally
+
+If your company is considering PureNix in order to tame a complicated Nix
+codebase, Jonas and I currently have time available consulting or freelance.
+Feel free to [get in touch](/about).  We are also available for any other
+Nix/Haskell/PureScript-related work.
+
+## Footnotes
+
+[^1]: This has been slightly simplified to make it a little easier to understand.
```
Comment (jonascarpay) [inline 2021-12-27-getting-started-with-purenix.md]:
```suggestion
[^1]: The formatting has been slightly cleaned up to make it a little easier to understand.
```

File: 2021-12-27-getting-started-with-purenix.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+```console
+$ nix repl ./output/Main/default.nix
+nix-repl> identity greeting
+"Hello, world!"
+nix-repl> fromMaybe "bye" (Just "hello")
+"hello"
+nix-repl> fromMaybe "bye" Nothing
+"bye"
+```
```
Comment (jonascarpay) [inline 2021-12-27-getting-started-with-purenix.md]: I keep being surprised by how nice this works out. Maybe we should make a screencast out of this or something

File: 2021-12-27-getting-started-with-purenix.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+Now that you have a taste of PureNix, the next section explains how to get
+started actually writing PureNix code.
+
+## Getting Started
+
+There is a post on the NixOS Discourse that explains how to
+[put together a development environment for PureNix](https://discourse.nixos.org/t/purenix-nix-backend-for-purescript/15756/3).
```
Comment (jonascarpay) [inline 2021-12-27-getting-started-with-purenix.md]: This may be worth inlining? Or making into a separate post in the series?

File: 2021-12-27-getting-started-with-purenix.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+Now that you have an idea how to get started with PureNix, what sorts of
+developers will find PureNix easy to use?
+
+## Who is PureNix for?
+
+If you have experience with PureScript, Nix, and Haskell, PureNix will be easy
+to get started with.  You may have to write some FFI for Nix builtins, but
+other than that you should have no trouble using PureNix.  It should feel
+quite natural.
```
Comment (jonascarpay) [inline 2021-12-27-getting-started-with-purenix.md]: I feel like this section is a little add odds with what this post seems to be -- a practical introduction to PureNix. This seems more at home in post 2? Or maybe it's worth moving post 2 up, as a non-technical introduction?

File: 2021-12-26-purenix.md (hunk 1 above)
Comment (cdepillabout) [reply 2021-12-26-purenix.md]: I was thinking that 2 (Why PureNix?) would basically just be the story of how we started writing PureNix:

- I wanted to write `cabal2nixWithoutIFD`
- I thought writing PureNix would be a funny way to do it
- I talked to you about it, and you had a couple different suggestions (like a Haskell backend based on core, or just a Haskell DSL).
- After looking around a bit, doing it with PureScript seemed like it would be the easiest.
- We got PureNix done in about a month.  It worked better than either of us expected.
- We got a couple libraries ported a couple months after that.  It was a little time-consuming, but it mostly worked out well.
- Writing `cabal2nixWithoutIFD` was relatively easy.
- Basically everything worked better than we had expected.

I thought of calling this post "The History of PureNix", but it feels weird to call it a "history" when it has only existed for a few months.  Although I agree that "Why PureNix" is confusing as well.

For 4 (PureNix vs. Nix), I was thinking of showing some expressions written in both Nix and PureNix.  I was planning on trying to make the argument that for simple derivations (like when you're using Nix effectively as a bash templating language), raw Nix is easier.  But when you're writing actual code, PureNix can be easier.

File: 2021-12-27-getting-started-with-purenix.md (hunk 2 above)
Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: Oh thanks!  That's much nicer!  I fixed this in 78c9dd8.

File: 2021-12-27-getting-started-with-purenix.md (hunk 3 above)
Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: Thanks, fixed in add2b8b.

File: 2021-12-27-getting-started-with-purenix.md (hunk 4 above)
Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: Yeah, we really should.  Aside from type-classes, it is super nice.

File: 2021-12-27-getting-started-with-purenix.md (hunk 5 above)
Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: Yeah, that's a good point.  I was basically doing the laziest thing I could here.

I thought about this a little bit, and I think what would be ideal is to have the content from https://discourse.nixos.org/t/purenix-nix-backend-for-purescript/15756/3 as an official document in the purenix repo.

I'll clean it up that post a bit and send it as a PR to the purenix repo.

File: 2021-12-27-getting-started-with-purenix.md (hunk 6 above)
Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: That's a good point.  Maybe I'll take this section and just make it its own post.  I could probably expand it a little as well.

Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: I removed this section in 80aa5e6.

Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: I re-added this section as a new post in 69a0505.

File: 2021-12-27-getting-started-with-purenix.md (hunk 5 above)
Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: I've sent a PR for this as https://github.com/purenix-org/purenix/pull/38.

Comment (cdepillabout) [reply 2021-12-27-getting-started-with-purenix.md]: Fixed in this blog post in 34c2a58.
